### DISCUSSION TITLE
error returned from database: (code: 5) database is locked

### DISCUSSION THREAD

**Discussion by BlockheadVFX-IT** [rustdesk/rustdesk-server-pro#271] [General]
Hi All, 
We've got this issue while logging in into Rust desk console. 

It says 
error returned from database: (code: 5) database is locked

We have restarted the server and tried few workarounds but the issues seems to persist. Can someone help?

Thanks

**Comment by rustdesk** [comment]
Could you stop the Pro server, find the db.sqlite3 file, and execute the `.recover` command in the SQLite CLI `sqlite3 db.sqlite3`? or run `sqlite3 db.sqlite3 .recover`;

or try out
`sqlite3 db.sqlite3 rollback`

**Reply by rustdesk** [reply]
Anything shown if you run `sqlite3 db.sqlite3 rollback`?

**Reply by BlockheadVFX-IT** [reply]
![image](https://github.com/rustdesk/rustdesk/assets/152861034/f7979977-738d-4f62-9930-45ca4bc37a4c)

**Reply by BlockheadVFX-IT** [reply]
i have had a another ssh session established where i stopped hbbs and hbbr , pls ignore service starting command just before i passed on rollback command

**Reply by rustdesk** [reply]
Is it possible you send the db files to us for debugging?

**Reply by BlockheadVFX-IT** [reply]
sure. how do i send it to you? just attach here ? or just send me email, i will send it to you

**Reply by rustdesk** [reply]
`tar czf db.tgz /var/lib/rustdesk-server`, you can send to info@rustdesk.com
BTW, what Linux distro and version are you running?

**Reply by rustdesk** [reply]
Could you confirm if there is only one `hbbs` process running? or there is a docker using the same db?

**Reply by fufesou** [reply]
I've found some links that may help. Pls backup the files before do changes.

1. Find the process that occupies the db file. `fuser <file>`
2. Show the lock status. https://www.beekeeperstudio.io/blog/how-to-solve-sqlite-database-is-locked-error#:~:text=%E2%80%9CPRAGMA%20lock_status%E2%80%9D
3. Consider create a backup of the database. These commands do not break my db file. https://www.arysontechnologies.com/blog/fix-sqlite-error-database-locked/#:~:text=To%20fix%20%E2%80%9CSQLite%20database%20is%20locked%20error%20code%205%E2%80%9D%20the%20best%20solution%20is%20to%20create%20a%20backup%20of%20the%20database

**Reply by fufesou** [reply]
`fuser` is also useful to find out if a process is using the file.

**Reply by BlockheadVFX-IT** [reply]
> `tar czf db.tgz /var/lib/rustdesk-server`, you can send to [info@rustdesk.com](mailto:info@rustdesk.com) BTW, what Linux distro and version are you running?

@rustdesk We're using Ubuntu 22.04 

root@akl08vrustdesk01:/home/akl08vrustdesk01# cat /etc/os-release 
PRETTY_NAME="Ubuntu 22.04.4 LTS"
NAME="Ubuntu"
VERSION_ID="22.04"
VERSION="22.04.4 LTS (Jammy Jellyfish)"
VERSION_CODENAME=jammy
ID=ubuntu
ID_LIKE=debian
HOME_URL="https://www.ubuntu.com/"
SUPPORT_URL="https://help.ubuntu.com/"
BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
UBUNTU_CODENAME=jammy

**Reply by BlockheadVFX-IT** [reply]
> Could you confirm if there is only one `hbbs` process running? or there is a docker using the same db?

We have hbbr and hbbs running

No docker containers being used in that server

**Reply by BlockheadVFX-IT** [reply]
> `tar czf db.tgz /var/lib/rustdesk-server`, you can send to [info@rustdesk.com](mailto:info@rustdesk.com) BTW, what Linux distro and version are you running?

sending it now , please expect in about 5-10 min 
thanks @rustdesk

**Reply by BlockheadVFX-IT** [reply]
sent

**Reply by rustdesk** [reply]
Received. Thank you very much. @fufesou  is debugging on this.

**Reply by BlockheadVFX-IT** [reply]
> I've found some links that may help. Pls backup the files before do changes.
> 
> 1. Find the process that occupies the db file. `fuser <file>`
> 2. Show the lock status. https://www.beekeeperstudio.io/blog/how-to-solve-sqlite-database-is-locked-error#:~:text=%E2%80%9CPRAGMA%20lock_status%E2%80%9D
> 3. Consider create a backup of the database. These commands do not break my db file. https://www.arysontechnologies.com/blog/fix-sqlite-error-database-locked/#:~:text=To%20fix%20%E2%80%9CSQLite%20database%20is%20locked%20error%20code%205%E2%80%9D%20the%20best%20solution%20is%20to%20create%20a%20backup%20of%20the%20database


Thank you. I did try this .backup thing but didnt make any difference.

**Reply by rustdesk** [reply]
We tested your files, so far, it is working well on our side. Could you please try out docker? Docker env is quite pure than your env. :(

https://rustdesk.com/docs/en/self-host/rustdesk-server-pro/installscript/docker/

You just need to move data to `./data`. 

`cp -rf /var/lib/rustdesk-server/* ./data/`

or

`ln -s /var/lib/rustdesk-server ./data`.

**Reply by BlockheadVFX-IT** [reply]
Thanks. i will do that now

**Reply by BlockheadVFX-IT** [reply]
![image](https://github.com/rustdesk/rustdesk-server-pro/assets/152861034/650e3794-a237-46eb-b147-763df2a0a32b)
database is still locked

**Reply by rustdesk** [reply]
Thanks, we will test, you do not need to send us password, we will reset it on our side.

**Reply by BlockheadVFX-IT** [reply]
@rustdesk 
oops , sorry i didn't get a chance to send it last week.

**Comment by fufesou** [comment]
Hi, can you please run the following commands? To see which api causes this error

```bash
$ sudo su
# systemctl stop rustdesk-hbbs && cd /var/lib/rustdesk-server
# RUST_LOG=debug hbbs > 1.log
```

**Reply by BlockheadVFX-IT** [reply]
[1.log](https://github.com/user-attachments/files/15948474/1.log)

**Reply by BlockheadVFX-IT** [reply]
The db lock error still persist. 

Specially when i try to login using Azure , also when i try to delete the old offline devices and also when i try to change the session expiry time. 

Its been quiet annoying. Should i consider redeploying server using docker instead?

**Reply by fufesou** [reply]
So the db is not always locked?
This error is triggered during some operations? 

I have noticed that there is a database operation that is extremely slow.

**Reply by fufesou** [reply]
Is the issue existed from the beginning? Or did it appear after a period of use?
Are there other programs running on your machine?

The database operations are really too slow sometimes.

**Reply by BlockheadVFX-IT** [reply]
It appeared after using for couple of months. No issues initially. 

No other programs are running except Wazuh agent which is being used for monitoring certain stuffs. Wazuh was there from day 1 , so i suspect its not the causing the issue.

Is there a way to migrate all the sqlite3 stuff to any other database ( like PostgreSQLor something)?

If i do the operation( like deleting old computer entry from rustdesk, it will throw some db lock issue and then later on after couple of tries, it deletes successfully)

**Reply by fufesou** [reply]
Sorry, migration to postgres is not possible at this time.

> If i do the operation( like deleting old computer entry from rustdesk, it will throw some db lock issue and then later on after couple of tries, it deletes successfully)

The log also show that the db is too slow sometimes.

Could you please check the CPU, memory, disk, etc. of the machine?

**Reply by BlockheadVFX-IT** [reply]
![image](https://github.com/rustdesk/rustdesk-server-pro/assets/152861034/d8bc7994-b3f4-423b-b84e-4954d80fcd57)

**Reply by fufesou** [reply]
Your machine resources appear to be sufficient.

I have no idea why it is slow.

I even found that inserting or updating `settings` could take up to 10 seconds. This table will not be locked at all.

It seems the reason is not the transaction lock.

```
summary="INSERT INTO settings (key, …" db.statement="\n\nINSERT INTO\n  settings (key, value)\nVALUES\n  ($1, $2) ON CONFLICT (key) DO\nUPDATE\nSET\n  value = EXCLUDED.value\n" rows_affected=0 rows_returned=0 elapsed=10.034284975s
```

Your database operation is sometimes slow and sometimes fast. It seems that something is affecting it.

**Reply by BlockheadVFX-IT** [reply]
let me do one more thing , I'll spin up a new VM , deploy hbbs container , copy keys , database and stuffs to new host and test it. If its to do with machine , then db wont get locked because of whatever reason

**Reply by rustdesk** [reply]
It seems more related to disk performance. hbbs has small cpu and memory usage.

we will try change `online` update to batch update, this may be more friendly to disk (less fsync called).

**Comment by BlockheadVFX-IT** [comment]
It must be something to do with the machine then? or maybe i'm trying to
update things too fast ?

**Reply by rustdesk** [reply]
You could try out this update, https://github.com/rustdesk/rustdesk-server-pro/releases/tag/1.3.8, we have done some improvement of disk write.

**Reply by BlockheadVFX-IT** [reply]
Well it was the storage issue , i suppose. I moved the same VM to different storage and did the exact same thing which was throwing the db locked issue. That issue is now gone.


also , i have now updated to latest one as well.

Thank you @rustdesk and @fufesou

**Reply by BlockheadVFX-IT** [reply]
@rustdesk quick question 
I have some ghost machine in my device list without ID's . Deleting them looks impossible. Is there any other way to force delete them? 

![image](https://github.com/rustdesk/rustdesk-server-pro/assets/152861034/d24ac2a3-2359-4c59-a0c5-8bf077ce1090)

**Reply by rustdesk** [reply]
@21pages

**Reply by BlockheadVFX-IT** [reply]
I just want to appreciate on how responsive you guys are. Thank you so so so much.

**Reply by 21pages** [reply]
@BlockheadVFX-IT 
I want to find how the name "-" happens . Do you remember what operations you did on the device alias "PLUTO"? Does this device still exist in the server device list? If not, how was it deleted? Was it deleted through the sql tool?

**Reply by rustdesk** [reply]
Please try out https://github.com/rustdesk/rustdesk-server-pro/releases/tag/1.3.9

**Reply by BlockheadVFX-IT** [reply]
okay sorry for the delay. But that thing happened when i Disabled the ID change from management console and manually went in and tried changing ID from end workstation. It was just a random incident from me.

**Reply by BlockheadVFX-IT** [reply]
how do i deploy the pre release thingy if i run as a package?

I have been using this script to update every single time 

 bash <(wget -qO- https://raw.githubusercontent.com/rustdesk/rustdesk-server-pro/main/update.sh)

**Reply by BlockheadVFX-IT** [reply]
nevermind, i have tweaked the script and updated it to 1.3.9 pre release version 
![image](https://github.com/rustdesk/rustdesk-server-pro/assets/152861034/0c98b37d-6ae7-40cb-bec3-1fe61b498f45)

**Reply by BlockheadVFX-IT** [reply]
good news, i can now delete the ghost devices
Thanks @rustdesk
